src/Graphql/Type/SchemaBuilder.php

```diff
@@ -209,55 +263,66 @@ private function convertType(Type $type, bool $isInput = false, bool $isMutation
         }
 
-        return $type->isNullable() || ($isMutation && 'update' === $mutationName) ? $graphqlType : GraphQLType::nonNull($graphqlType);
+        return $type->isNullable() || (null !== $mutationName && 'update' === $mutationName) ? $graphqlType : GraphQLType::nonNull($graphqlType);
     }
 
     /**
      * Gets the object type of the given resource.
      *
      * @return ObjectType|InputObjectType
      */
-    private function getResourceObjectType(string $resource, ResourceMetadata $resourceMetadata, bool $isInput = false, bool $isMutation = false, string $mutationName = null)
+    private function getResourceObjectType(string $resourceClass, ResourceMetadata $resourceMetadata, bool $input = false, string $mutationName = null): GraphQLType
     {
-        $shortName = $resourceMetadata->getShortName().($isInput ? 'Input' : '').($isMutation ? ucfirst($mutationName).'Mutation' : '');
+        $shortName = $resourceMetadata->getShortName();
+        if ($input) {
+            $shortName .= 'Input';
+        }
+        if (null !== $mutationName) {
+            $shortName .= ucfirst($mutationName).'Mutation';
+        }
 
-        if (isset($this->resourceTypesCache[$shortName])) {
-            return $this->resourceTypesCache[$shortName];
+        if (isset($this->graphqlTypes[$resourceClass][$mutationName][$input])) {
+            return $this->graphqlTypes[$resourceClass][$mutationName][$input];
         }
 
         $configuration = [
             'name' => $shortName,
             'description' => $resourceMetadata->getDescription(),
             'resolveField' => $this->defaultFieldResolver,
-            'fields' => function () use ($resource, $isInput, $isMutation, $mutationName) {
-                return $this->getResourceObjectTypeFields($resource, $isInput, $isMutation, $mutationName);
+            'fields' => function () use ($resourceClass, $input, $mutationName) {
+                return $this->getResourceObjectTypeFields($resourceClass, $input, $mutationName);
             },
+            'interfaces' => [$this->getNodeInterface()],
         ];
 
-        return $this->resourceTypesCache[$shortName] = $isInput ? new InputObjectType($configuration) : new ObjectType($configuration);
+        return $this->graphqlTypes[$resourceClass][$mutationName][$input] = $input ? new InputObjectType($configuration) : new ObjectType($configuration);
     }
 
     /**
      * Gets the fields of the type of the given resource.
      */
-    private function getResourceObjectTypeFields(string $resource, bool $isInput = false, bool $isMutation = false, string $mutationName = null): array
+    private function getResourceObjectTypeFields(string $resource, bool $input = false, string $mutationName = null): array
     {
         $fields = [];
         $idField = ['type' => GraphQLType::id()];
```

Comment:
This should be wrapped by GraphQLType::nonNull() as it implements Relay Node interface now, and Node.id required to be nonNull. Commenting here as I'm not sure how much you have progressed after this commit, will submit PR if required. Thank you for all your work!

Reply:
This has not be fixed. Can I let you do the PR? Thanks for the review!

Reply:
@dunglas saw you already did that, though it would be faster for you anyway, as you were still working in that file :) Thank you for all your work once again! Waiting for your blog entry on GraphQL :) have a good weekend 🎉